Kill the "shadow-functions.js" internal plugin in favor of an explicit helper

loganfsmyth opened the pull request:
| Q                        | A <!--(yes/no) -->
| ------------------------ | ---
| Patch: Bug Fix?          | Y
| Major: Breaking Change?  | Y, if something was relying on this stuff not being unwrapped immediately, or was using the `_forceShadow`, `.shadow` or `._shadowedFunctionLiteral` properties. Hopefully minimal though.
| Minor: New Feature?      | Maybe?
| Deprecations?            | 
| Spec Compliancy?         | 
| Tests Added/Pass?        | 
| Fixed Tickets            | Fixes #5597, Fixes #3930 <!-- rm the quotes to link the issues -->
| License                  | MIT
| Doc PR                   | <!-- if yes, add `[skip ci]` to your commit message to skip CI builds -->
| Dependency Changes       | 

This removes the private "shadow"-related properties from the AST and relies on explicit traversal to handle the conversion of arrow functions into non-arrow functions.

I've also added support for handling `super` and `new.target` into the generic utility to handle cases that we could not handle properly with the old implementation.

This also has the added benefit of not performing unexpected transformations on the AST when no plugins were enabled by the user, since we've had users reporting this as an annoyance on several occasions. This is also a bit of a blocker for any work trying to make Babel a more general codemod-like tool.

loganfsmyth commented on packages/babel-plugin-transform-async-generator-functions/test/fixtures/nested/arrows-in-declaration/expected.js:
```diff
@@ -2,9 +2,7 @@ let g = (() => {
-    (function () {
-      return _this;
-    });
+    () => this;
```
These changes are from the removed call on https://github.com/babel/babel/pull/5677/files#diff-3953cdd7a44c58e0431a2e202efb64d4L31 that wasn't needed.

loganfsmyth commented on packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/object-method-with-super/expected.js:
```diff
@@ -0,0 +1,11 @@
+class Foo extends class {} {
+  method() {
+    var _superprop_callMethod = (..._args) => super.method(..._args);
```
It's not pretty, but it's the only way we can deal with this.

xtuc replied:
This looks okay to me. Nice job, you use the same reference to the `super.method` call (instead of two separate) :+1:

loganfsmyth commented on packages/babel-plugin-transform-es2015-block-scoping/src/index.js:
```diff
@@ -416,56 +416,79 @@ class BlockScoping {
     // build the closure that we're going to wrap the block with, possible wrapping switch(){}
     const fn = t.functionExpression(null, params,
       t.blockStatement(isSwitch ? [block] : block.body));
-    fn.shadow = true;
```
This whole set of changes is gross and probably the most likely to be missing edge cases, but it works as far as I can tell.

These changes were all needed so that we can get access to the final NodePath for the function, since it wasn't available anywhere before.

loganfsmyth commented on packages/babel-plugin-transform-es2015-block-scoping/test/fixtures/general/for-break-continue-return/expected.js:
```diff
@@ -1,5 +1,5 @@
 (function () {
-  var _loop = function (i) {
+  var _loop2 = function (i) {
```
A change in ordering in block scoping means these variable names are generated in opposite order now, thus breaking the test.

loganfsmyth commented on packages/babel-plugin-transform-es2015-classes/src/index.js:
```diff
@@ -46,6 +46,10 @@ export default function ({ types: t }) {
         path.replaceWith(new Constructor(path, state.file).run());
+
+        if (path.isCallExpression() && path.get("callee").isArrowFunctionExpression()) {
+          path.get("callee").arrowFunctionToExpression();
```
We don't really have a firm policy for when it's cool to insert actual arrows, so I added a few new arrows in place of `.shadow` but explicitly converted them to keep the tests passing.

hzoo replied:
Would hope that the dependency system would allow us to add/remove, preferring to instead arrows/newer syntax

loganfsmyth replied:
Yeah, we can always drop this now if we want, since the plugins are already in a fine order for it AFAIK, I just figured it would be better to leave that for a separate PR.

loganfsmyth commented on packages/babel-plugin-transform-es2015-object-super/src/index.js:
```diff
@@ -1,46 +1,39 @@
+        const getObjectRef = () => objectRef = objectRef || path.scope.generateUidIdentifier("obj");
 
-      ObjectExpression: {
-        exit(path, file) {
```
This implementation had to change because the usage of the `exit` handler meant that the arrow transform would fail for
```
var obj = {
  method(){
    var fn = () => super.method();
  }
};
```
since it can't handle `super`, meaning that for object-super to work, it needs to run before the arrow transform.

hzoo replied:
So there's no need for the `CONTAINS_SUPER` check anymore?

loganfsmyth replied:
I don't think there's a good reason to have it if we've not using the `exit` handler, since the main benefit of it was avoiding traversing the methods more than once. and there's no way to avoid that without the `exit`, but the `exit` would run it all in the wrong order.

loganfsmyth commented on packages/babel-traverse/src/path/replacement.js:
```diff
@@ -239,6 +238,8 @@ export function replaceExpressionWithStatements(nodes: Array<Object>) {
       }
     }
 
+    this.get("callee").arrowFunctionToExpression();
```
Again just to make the tests pass, we could also leave this out and insert real arrows if we want to.

loganfsmyth commented on packages/babel-types/src/definitions/es2015.js:
```diff
@@ -245,10 +245,10 @@ defineType("MetaProperty", {
     // todo: limit to new.target
     meta: {
-      validate: assertValueType("string"),
+      validate: assertNodeType("Identifier"),
```
These were just wrong :P

hzoo commented on packages/babel-plugin-transform-es2015-arrow-functions/test/fixtures/arrow-functions/spec/expected.js:
```diff
@@ -14,7 +14,7 @@ function foo() {
 
     return function () {
       babelHelpers.newArrowCheck(this, _this2);
-      return this;
```
looks like this used to be a bug too?

loganfsmyth replied:
Since it is `.bind` both work, this way is just a little more consistent with the non-spec arrow transform, since they share more code now.

Kovensky replied:
I'd prefer if we could keep `this` as `this`, but it's not like it's incorrect code here. Minifies better too!

loganfsmyth replied:
Up for expanding on your reasoning? My motivation was mostly that it's less separation between spec and non-spec mode, but it could be made to work either way I guess.

loganfsmyth replied:
Alright, I added support for spec arrow in subclass constructors, and made `this` stay `this` in spec arrows except for the subclass constructor case.

xtuc commented on packages/babel-traverse/src/path/ancestry.js:
```diff
@@ -200,59 +200,3 @@ export function inType() {
- * Please Note, these flags are for private internal use only and should be avoided.
- * Only "shadow" is a public property that other transforms may manipulate.
- */
-
```
Could we keep this comment. It helped me understand what `shadow`'ing meant.

loganfsmyth replied:
Happy to chat if you have suggestions. This PR mostly removes the whole concept of "shadowing" that we had before, in favor of fixing everything up right away, so it's not obvious where we'd put this, or if it would be useful to people at this point.

xtuc replied:
Yes, I you did great by removing unnecessary complexity. The term variable shadowing as explained in the [wiki there](https://en.wikipedia.org/wiki/Variable_shadowing) is confusing to me. It seems that shadowing wasn't the correct concept for passing `this` since it's just a transpilation technique. We could maybe chat about that on Slack if you have time?

loganfsmyth replied:
Yeaaah talk about confusing naming.

xtuc replied:
I talked with @loganfsmyth and now I'm fine with that.

Kovensky commented on packages/babel-traverse/src/path/conversion.js:
```diff
@@ -25,14 +25,415 @@ export function ensureBlock() {
+        t.thisExpression(),
+        t.identifier(thisBinding),
+      ]))
+    );
```
Looks like I got you a merge conflict by merging #5620

loganfsmyth replied:
No worries.

Kovensky commented on packages/babel-traverse/src/path/conversion.js:
```diff
@@ -25,14 +25,415 @@ export function ensureBlock() {
+    ThisExpression(child) {
+      thisPaths.push(child);
+    },
+    JSXIdentifier(child) {
```
Is this for `<this.Foo/>`?

loganfsmyth replied:
It is, I guess I should make this `JSXMemberExpression` now that you mention it though.

loganfsmyth replied:
Added some checks to explicitly verify the location of the identifier.

Kovensky commented on packages/babel-traverse/test/arrow-transform.js:
```diff
@@ -0,0 +1,439 @@
+      () => super();
+      () => this;
+    `, `
+      var _supercall = (..._args) => _this = super(..._args),
```
Was wondering if something would catch multiple `super()` invocations, but I guess that's the job of the JS engine / class transform.

loganfsmyth replied:
That I think we should leave up to the class transform, it could certainly add logic for that.

loganfsmyth commented on packages/babel-helper-remap-async-to-generator/src/index.js:
```diff
@@ -95,6 +90,9 @@ function classOrObjectMethod(path: NodePath, callId: Object) {
   node.generator = false;
+
+  // Unwrap the wrapper IIFE's environment so super and this and such still work.
+  path.get("body.body.0.argument.callee.arguments.0").unwrapFunctionEnvironment();
```
It's gross garbage :P It's reaching in to the path where we put the IIFE, since we only get access to the path _after_ it is inserted, so we have to insert it then do this gross stuff to get the reference.

loganfsmyth commented on packages/babel-plugin-transform-async-to-generator/test/fixtures/async-to-generator/deeply-nested-asyncs/expected.js:
```diff
@@ -11,11 +11,12 @@ let s = (() => {
               innerArgs[_key2 - 2] = arguments[_key2];
             }
 
-            yield z;
```
I forget, but this change seems _more_ like what it's supposed to do, so I didn't worry about it. It's possible I looked into the reason, but I don't remember anymore.

loganfsmyth commented on packages/babel-plugin-transform-es2015-arrow-functions/src/index.js:
```diff
@@ -1,34 +1,17 @@
-        }
+        path.arrowFunctionToExpression({
+          // While other utils may be fine inserting other arrows to make more transforms possible,
+          // the arrow transform itself absolutely cannot insert new arrow functions.
```
Why not both? :P

loganfsmyth commented on packages/babel-plugin-transform-es2015-block-scoping/src/index.js:
```diff
@@ -416,56 +416,79 @@ class BlockScoping {
-        t.variableDeclarator(ref, fn),
-      ]));
-    }
+    const has = this.has;
```
I think this was leftover from copy-pasting from https://github.com/babel/babel/pull/5677/files/c8ff06f90661f5e67a93b7e2f2d387a029edd98a#diff-775ae6452de339480d54604c7d217b5dL463

Looks like `this.has` is used one other place, but otherwise isn't updated, so you're right I could move this up farther.

loganfsmyth commented on packages/babel-plugin-transform-es2015-block-scoping/src/index.js:
```diff
@@ -416,56 +416,79 @@ class BlockScoping {
       fn.async = true;
       call = t.awaitExpression(call);
+      basePath = ".argument" + basePath;
     }
```
No idea, I didn't realize this logic even did this until inspecting this code.

loganfsmyth commented:
Okay, I rebased to fix the merge conflict with spec function naming, and addressed the few issues people have raised except the one question of `this` vs `_this` in arrow spec mode. Happy to chat on that on Slack tomorrow.

mikemorancodes commented:
What if I just need to fix the #3930 issue?
